gusto/linear_solvers.py

```diff
@@ -221,6 +222,214 @@ def solve(self):
+        # component of the gradient
+
+        # the pi prime term (here, bars are for mean and no bars are
+        # for linear perturbations)
+
+        pi = pibar_theta*theta + pibar_rho*rho
+
+        # vertical projection
+        def V(u):
+            return k*inner(u,k)
+
+        # specify degree for some terms as estimated degree is too large
+        dxp = dx(degree=(self.quadrature_degree))
+        dS_vp = dS_v(degree=(self.quadrature_degree))
+        dS_hp = dS_h(degree=(self.quadrature_degree))
+        ds_vp = ds_v(degree=(self.quadrature_degree))
+        ds_tbp = ds_t(degree=(self.quadrature_degree)) + ds_b(degree=(self.quadrature_degree))
+
+        rhobar_tr = Function(Vtrace)
+        rbareqn = (l0('+') - avg(rhobar))*dl('+')*(dS_vp + dS_hp) + \
+                  (l0 - rhobar)*dl*ds_vp + \
+                  (l0 - rhobar)*dl*ds_tbp
+        rhobar_prob = LinearVariationalProblem(lhs(rbareqn),rhs(rbareqn),rhobar_tr)
+        self.rhobar_solver = LinearVariationalSolver(rhobar_prob,
+                                                     solver_parameters={'ksp_type':'preonly',
+                                                                        'pc_type':'bjacobi',
+                                                                        'pc_sub_type':'lu'})
+
+        Aeqn = (
+            inner(w, (u - u_in))*dx
+            - beta*cp*div(theta*V(w))*pibar*dxp
+            - beta*cp*div(thetabar*w)*pi*dxp
+            + (phi*(rho - rho_in) - beta*inner(grad(phi), u)*rhobar)*dx
+            + beta*inner(phi*u, n)*rhobar_tr*(dS_v + dS_h)
+        )
+        if mu is not None:
+            Aeqn += dt*mu*inner(w,k)*inner(u,k)*dx
+        Aop = Tensor(lhs(Aeqn))
+        Arhs = rhs(Aeqn)
+
+        #  (A K)(U) = (U_r)
+        #  (L 0)(l)   (0  )
+
+        dl = dl('+')
+        l0 = l0('+')
+
+        K = Tensor(beta*cp*inner(thetabar*w,n)*l0*(dS_vp + dS_hp)
+                   + beta*cp*inner(thetabar*w,n)*l0*ds_vp
+                   + beta*cp*inner(thetabar*w,n)*l0*ds_tbp
+        )
+        L = Tensor(dl*inner(u,n)*(dS_vp + dS_hp)
+                   + dl*inner(u,n)*ds_vp 
+                   + dl*inner(u,n)*ds_tbp
+        )
+
+        #  U = A^{-1}(-Kl + U_r), 0=LU=-(LA^{-1}K)l + LA^{-1}U_r, so (LA^{-1}K)l = LA^{-1}U_r
+        # reduced eqns for l0
+        S = assemble(L * Aop.inv * K)
+        self.Rexp = L * Aop.inv * Tensor(Arhs)
+        # place to put the RHS (self.Rexp gets assembled in here)
+        self.R = Function(M)
+
+        # Set up the LinearSolver for the system of Lagrange multipliers
+        self.lSolver = LinearSolver(S, solver_parameters=self.params)
+        # a place to keep the solution
+        self.lambdar = Function(Vtrace)
+
+        # Place to put result of u rho solver
+        self.urho = Function(M)
+
+        # Reconstruction of broken u and rho
+        self.ASolver = LinearSolver(assemble(Aop),
+                                    solver_parameters={'ksp_type':'preonly',
```

Comment:
Here you want `lhs(Aeqn)` to perform the global reconstruction without Slate. However, you will need to modify `Aeqn` to include the jump terms as you normally would.

Probably better to just store all the cell integrals in something, say `A_dx`, and the facet terms in `A_dS` (with Slate-restricted syntax) and `A_jump_dS` (including jumps).

Then `Aop` above becomes `Aop = Tensor(lhs(A_dx + A_dS))` and the global reconstruct here would be `assemble(lhs(A_dx + A_jump_dS), ...)`. Just a thought.